src/tests/test_switch_map.cpp

```diff
@@ -0,0 +1,103 @@
+
+#include "mock_observer.hpp"
+
+SCENARIO("switch_map acts like flat_map except it unsubscribes the previous source when new source emits", "[flat_map]")
+{
+    auto mock = mock_observer<int>{};
+    GIVEN("observable of items")
+    {
+        auto obs = rpp::source::just(1,2,3);
+        WHEN("subscribe on it via switch_map")
+        {
+            obs.switch_map([](int v){return rpp::source::just(v, v*10);})
+                .subscribe(mock);
+            THEN("subscriber obtains values from observables obtained via switch_map")
+            {
+                CHECK(mock.get_received_values() == std::vector{1,10,2,20,3,30});
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        WHEN("subscribe on it via switch_map with error")
+        {
+            obs.switch_map([](int){return rpp::source::error<int>(std::make_exception_ptr(std::runtime_error{""}));})
+                .subscribe(mock);
+            THEN("subscriber obtains values from observables obtained via switch_map")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 1);
+            }
+        }
+        WHEN("subscribe on it via switch_map with empty")
+        {
+            obs.switch_map([](int){return rpp::source::empty<int>();})
+                .subscribe(mock);
+            THEN("subscriber obtains values from observables obtained via switch_map")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 1);
+                CHECK(mock.get_on_error_count() == 0);
+            }
+        }
+        WHEN("subscribe on it via switch_map with never")
+        {
+            auto sub = obs.switch_map([](int){return rpp::source::never<int>();})
+                .subscribe(mock);
+            THEN("subscriber obtains values from observables obtained via switch_map")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 0);
+                CHECK(sub.is_subscribed());
+            }
+        }
+        WHEN("subscribe on it via switch_map with mix of never + observable")
+        {
+            auto sub = obs.switch_map([](int){return rpp::source::never<int>();})
+                    .subscribe(mock);
+            THEN("subscriber obtains values from observables obtained via switch_map")
+            {
+                CHECK(mock.get_total_on_next_count() == 0);
+                CHECK(mock.get_on_completed_count() == 0);
+                CHECK(mock.get_on_error_count() == 0);
+                CHECK(sub.is_subscribed());
+            }
+        }
+        WHEN("subscribe on it via switch_map with never + observables")
```

Comment:
Can you add same test, but when never is last and no `on_completed` call as a result?

Reply:
That's a very good point. I'll add it.